terraform/transform_destroy_edge.go

```diff
@@ -163,46 +164,146 @@ func (t *DestroyEdgeTransformer) Transform(g *Graph) error {
 
-		if nonDestroyInstanceFound {
-			continue
+	for _, mod := range moduleMap {
+		modules = append(modules, mod)
+	}
+
+	// Sort them by path length, longest first, so that start with the deepest
+	// modules.  The order of modules at the same tree level doesn't matter, we
+	// just need to ensure that child modules are processed before parent
+	// modules.
+	sort.Slice(modules, func(i, j int) bool {
+		return len(modules[i].addr) > len(modules[j].addr)
+	})
+
+	for _, mod := range modules {
+		mod.removeUnused(g)
+	}
+
+	return nil
+}
+
+// pruneUnusedNodesMod is a container to hold the nodes that belong to a
+// particular configuration module for the pruneUnusedNodesTransformer
+type pruneUnusedNodesMod struct {
+	addr   addrs.Module
+	nodes  []dag.Vertex
+	closer *nodeCloseModule
+}
+
+// Remove any unused locals, variables, outputs and expanders.  Since module
+// closers can also lookup expansion info to detect orphaned instances, disable
+// them if their associated expander is removed.
+func (m *pruneUnusedNodesMod) removeUnused(g *Graph) {
+	// We modify the nodes slice during processing here.
+	// Make a copy so no one is surprised by this changing in the future.
+	nodes := make([]dag.Vertex, len(m.nodes))
+	copy(nodes, m.nodes)
+
+	// since we have no defined structure within the module, just cycle through
+	// the nodes in each module until there are no more removals
+	removed := true
+	for {
+		if !removed {
+			return
 		}
+		removed = false
+
+		for i := 0; i < len(nodes); i++ {
+			// run this in a closure, so we can return early rather than
+			// dealing with complex looping and labels
+			func() {
+				n := nodes[i]
+				switch n.(type) {
+				case graphNodeTemporaryValue:
+					// temporary value, which consist of variables, locals, and
+					// outputs, must be kept if anything refers to them.
+					if n, ok := n.(GraphNodeModulePath); ok {
+						// root outputs always have an implicit dependency on
+						// remote state.
+						if n.ModulePath().IsRoot() {
+							return
+						}
+					}
+					for _, vv := range g.UpEdges(n) {
+						// keep any value which is connected through a
+						// reference
+						if _, ok := vv.(GraphNodeReferencer); ok {
+							return
+						}
+					}
+
+				case graphNodeExpandsInstances:
+					// Any nodes that expand instances are kept when their
+					// instances may need to be evaluated.
+					for _, vv := range g.UpEdges(n) {
+						switch vv.(type) {
+						case graphNodeExpandsInstances:
+							// expanders can always depend on module expansion
+							// themselves
+							return
+						case GraphNodeResourceInstance:
+							// resource instances always depend on their
+							// resource node, which is an expander
+							return
+						}
+					}
+
+				default:
+					return
+				}
 
-		// connect all the through-edges, then delete the node
-		for _, d := range g.DownEdges(n) {
-			for _, u := range g.UpEdges(n) {
-				g.Connect(dag.BasicEdge(u, d))
-			}
+				g.Remove(n)
```

Comment:
You're dropping the log message in this section -- ok with that?

Reply:
Yeah, that was intentional, but after some more testing I have a better idea of what messages I want to see from this, and will add them in with the additional patches